File: tests/go/samples/after.go
// Package example demonstrates basic Go structures
package example

import (
	"fmt"
	"time"
)

// Config holds application configuration
type Config struct {
	Host string
	Port int
}

// Logger is an interface for logging operations
type Logger interface {
	Log(message string) error
	LogError(err error)
}

// DatabaseManager manages database connections and queries
type DatabaseManager struct {
	connectionString string
	connected        bool
}

// NewDatabaseManager creates a new DatabaseManager instance
func NewDatabaseManager(connStr string) *DatabaseManager {
	return &DatabaseManager{
		connectionString: connStr,
		connected:        false,
	}
}

// Connect establishes database connection
func (db *DatabaseManager) Connect() error {
	fmt.Println("Connecting to database")
	db.connected = true
	return nil
}

// Health is a liveness probe
func Health() bool {
	return true
}

// Query executes a SQL query and returns results
func (db *DatabaseManager) Query(sql string) ([]map[string]interface{}, error) {
	return nil, nil
}

// UserService handles user-related operations
type UserService struct {
	db *DatabaseManager
}

// CreateUser creates a new user in the database
func (s *UserService) CreateUser(username, email string) (int, error) {
	return 1, nil
}

// GetUser retrieves user by ID
func (s *UserService) GetUser(userID int) (map[string]interface{}, error) {
	return nil, nil
}

// DeleteUser deletes a user by ID
func (s *UserService) RemoveUser(userID int) error {
	return nil
}

// ValidateEmail validates email format
func ValidateEmail(email string, strict bool) bool {
	return len(email) > 0 && email[0] != '@'
}

// FormatTimestamp formats a timestamp to string
func FormatTimestamp(t time.Time) string {
	return t.Format("2006-01-02 15:04:05")
}

// main is the entry point
func main() {
	db := NewDatabaseManager("postgresql://localhost/mydb")
	retryConnect(db)
	fmt.Println("Application started")
}

// MaxRetries is the retry budget for connection attempts
const MaxRetries = 5

// retryConnect retries the connection until it succeeds or the budget is spent (internal)
func retryConnect(db *DatabaseManager) error {
	var err error
	for attempt := 0; attempt < MaxRetries; attempt++ {
		if err = db.Connect(); err == nil {
			return nil
		}
	}
	return err
}

File: tests/ruby/samples/after.rb
# Example Ruby file for testing the scanner

require 'json'
require 'net/http'
require_relative 'helpers'

# Data access layer module
module DataAccess
  # Manages database connections and queries
  class DatabaseManager
    attr_accessor :connection_string
    attr_reader :connection

    def initialize(connection_string)
      @connection_string = connection_string
      @connection = nil
    end

    # Establish database connection
    def connect
      puts "Connecting to #{@connection_string}"
    end

    # Execute a SQL query
    def query(sql)
      []
    end

    # Class method for creating a default instance
    def self.create_default
      new("postgresql://localhost/mydb")
    end
  end
end

# Handles user-related operations
class UserService
  def initialize(db)
    @db = db
  end

  # Create a new user
  def create_user(username, email)
    validate_email(email) ? 1 : 0
  end

  # Retrieve user by ID
  def get_user(user_id)
    nil
  end

  # Delete a user
  def remove_user(user_id)
    true
  end
end

# Validate email format
def validate_email(email, strict = false)
  email.include?('@')
end

# Liveness probe
def health
  { status: 'ok' }
end

# Default number of connection attempts
DEFAULT_ATTEMPTS = 5

# Retries a block a bounded number of times
class Retrier
  def initialize(attempts = DEFAULT_ATTEMPTS)
    @attempts = attempts
  end

  # Run the block until it returns true or the attempts are spent
  def run(&action)
    @attempts.times { return true if attempt(&action) }
    false
  end

  private

  # Execute one attempt (internal)
  def attempt
    yield == true
  end
end

# Main entry point
def main
  db = DataAccess::DatabaseManager.create_default
  service = UserService.new(db)
  Retrier.new.run { db.connect }
  puts "Application started"
end

main if __FILE__ == $PROGRAM_NAME
